setwd("/mnt/data/BMOHAMED/TRAX_nov2019/tRAX-master/nomultimap_FedvsStarved/counts_analysis/tRNAsonly/apeglm_minCount10/")

library(DESeq2)
library(tidyverse)
library(dplyr)

#################################################################################################################################

# LOADING OUR DATASETS

sample_information <- read_tsv("sampleinformation.txt")

seq_data <- read_tsv("/mnt/data/BMOHAMED/TRAX_nov2019/tRAX-master/nomultimap_FedvsStarved/counts_analysis/tRNAsonly/apeglm_minCount10/raw_counts_tRNAsonly.txt")

# Filter out all rows that have any column < 10

seq_data <- seq_data[apply(seq_data[, -1], MARGIN = 1, function(x) all(x > 10)), ]

newcolnames <- c("Gene", "Fed1", "Fed2", "Fed5", "Starved1", "Starved2", "Starved5")
colnames(seq_data) <- newcolnames

countdata <- seq_data %>%
  column_to_rownames("Gene") %>%
  as.matrix()

# Loaded the sample file containing (1) Bam file name (2) Sample name and (3) Condition
# Loaded the featureCounts.txt file generated by the counting step 
# comment = "#" comments out everything in the document accept the table itself 

#################################################################################################################################

# QUALITY ASSESMENT

# We can plot how many reads we have for each sample
# Although normalisation can account for imbalance, extreme differences may be indicitive of underlying problems 

library_sizes <- colSums(countdata)

conditionCol <- match(sample_information$Condition, c("Fed", "Starved"))

png("library_sizes_barplot.png")

barplot(library_sizes,
        names=names(library_sizes),
        las=1,
        main = "Barplot of Library sizes",
        col = conditionCol)
#abline(h=15e7, lty=2)

dev.off()

# In the above code, we're taking the sum of all columns and assigning it the variable library_sizes
# We then use the R barplot function to plot a barplot of all the library sizes in each condition 
# names tells us what to call each bar
# las details the orientation of the x-axis label ... 1 = horrizontal 2= verticle
# main is the title of the plot 
# abline is the dotted line ... h = y-intercept lty= how dotted the line will be with 1=solid line 
# 14million was picked and is an arbitary number 


#################################################################################################################################

# COUNT DISTRIBUTION BOXPLOTS 

# Count data is not normally distributed and so to examine raw counts, its best to log transfrom them (log2)
# Count data may contain 0's and so to avoid log2(0), we add 1 to every count and prevent errors 

logcounts <- log2(countdata + 1)

# getting the log2 counts 

# Check the distribuion of the counts and colour by condition 

conditionCol <- match(sample_information$Condition, c("Fed", "Starved")) + 1

# above we have created a colour vector 

png("counts_distribution_boxplot.png")

boxplot(logcounts,
        xlab= "",
        ylab= "Log2(Counts)",
        las = 1,
        col = conditionCol,
        main = "Counts Distribution Boxplot")
abline(h = median(as.matrix(logcounts)), col = "Blue")

dev.off()

# Above, we built a boxplot of the Log2(counts), coloured by condition
# The line is the median of the log2(counts) 
# remember we have to convert the log2(counts) into a matrix, hence the as.matrix
# Line is coloured blue 


#################################################################################################################################

# PRINCIPLE COMPONENT ANALYSIS - PRE DIFFERENTIAL EXPRESSION (SANITY CHECK)

library(ggfortify)

png("PreDifferentialExpression_PCA.png")

rlogcounts <- rlog(countdata)

pcDat <- prcomp(t(rlogcounts))

autoplot(pcDat,
         data = sample_information,
         colour = "Condition",
         shape = FALSE,
         size = 5)

dev.off()

#################################################################################################################################

# CONVERT COUNTS INTO A DESeqDataSet OBJECT

# A DESeqDataSet object is used by DESeq2 to store count data 
# It has a number of slots for storing; counts, sample information, the model design etc for DE

library(DESeq2)

all(sample_information$Sample == colnames(countdata))

design <- as.formula(~ batch + Condition) # correcting for batch effect by including in my design - note: always do condition of interest last  

model_matrix <- model.matrix(design, data = sample_information) 

ddsObj <- DESeqDataSetFromMatrix(countData = countdata,
                                 colData = sample_information,
                                 design = design)

ddsObj <- estimateSizeFactors(ddsObj)

sizeFactors <- sizeFactors(ddsObj)

ddsObj_results <- DESeq(ddsObj)

resultsNames(ddsObj_results)                                            

# FedvsStarved <- results(ddsObj_results, contrast = c("Condition", "Starved", "Fed"), alpha = 0.05)

library(apeglm)

FedvsStarved <- lfcShrink(ddsObj_results, coef = "Condition_Starved_vs_Fed", type = "apeglm")

#################################################################################################################################

# PCA post batch correction 

ddsObj$batch <- factor(rep(c("batch1", "batch2", "batch3")))

png("PCAplot_pre_batchcorrection.png")

vsd <- varianceStabilizingTransformation(ddsObj) # do a variance stablizing transformation on the model matrix 

plotPCA(vsd, "batch") +
  geom_text(aes(label = ddsObj$Sample), vjust = 1.5)+
  ggtitle("Fed vs Starved PCA (pre batch correction)")  # this should give you the variance associated with the effect of the batch# this should give you the variance associated with the effect of the batch 

dev.off()

png("PCAplot_post_batchcorrection.png")

assay(vsd) <- limma::removeBatchEffect(assay(vsd), vsd$batch) # applying limma's removeBatchEffect 

plotPCA(vsd, "batch") +
  geom_text(aes(label = ddsObj$Sample), vjust = 1.5)+
  ggtitle("Fed vs Starved PCA (post batch correction)")+
  coord_fixed(ratio = 1.3)  # this should give you the variance associated with the effect of the batch # plotting the residuals 

dev.off()

# Replot the PCAs in ggplot 

# need to find way to replot the PCA before and after batch correction 

#################################################################################################################################

#prep data for barplot input 

# Generate a table from the DESeq2 results 

FedvsStarved_table <- as.data.frame(FedvsStarved) %>%
  rownames_to_column("Gene") %>%
  rename(log2FC = log2FoldChange, FDR = padj)

# save the DESeq2 results table for future reference 

write_tsv(FedvsStarved_table, "/mnt/data/BMOHAMED/TRAX_nov2019/tRAX-master/nomultimap_FedvsStarved/counts_analysis/tRNAsonly/apeglm_minCount10/rawcounts_full3prime_combined_all_DESeq2results.txt")

# reformatt the table so it can be used for tRNA barplot generation 

# step 1 ... remove all rows that are non-tRNAs

barplot_input <- filter(FedvsStarved_table, grepl("tRNA-", Gene)) # filtering rows with tRNA- in the rowname

# step 2 ... convert Gene column into individual AA and anticodon columns etc  

barplot_input <- mutate(barplot_input, Label = Gene) # make a new column called Label 

barplot_input <- separate(barplot_input, Gene, sep = "-", into = c("RNA", "Isodecoder", "Anticodon", "Loci")) # seperate the rowname into columns

# step 3 ... remove all rows with NA in the FDR column 

barplot_input <- barplot_input[!is.na(barplot_input$FDR),]

#################################################################################################################################
# Bar plot of tRNA log fold changes 
#################################################################################################################################

library(ggplot2)
library(grid)
library(gridExtra)

df <- barplot_input

plot_list <- list()

for (iso in c("Ala", "Arg", "Asn", "Asp", "Cys", "Gln", "Glu", "Gly", "His", "Ile", "Leu", "Lys", "Met", "Phe", "Pro", "SeC", "Ser", "Thr", "Trp", "Tyr", "Val", "iMet")) {
  
  df2 <- subset(df, Isodecoder == iso) 
  
  temp_plot <- ggplot(data = df2, aes(x = Label, y = log2FC, fill = Anticodon, label = ifelse(FDR < 0.05, "*", "ns")))+
    geom_bar(stat = "identity")+
    geom_text(vjust = 0, nudge_y = ifelse(df2$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(df2$FDR < 0.05, 6, 4))+
    xlab("Isodecoder")+
    ylab(expression(paste(Log[2],"FC",sep="")))+
    ylim(c(-5, 5))+
    theme_bw()+
    theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
    ggtitle(iso, "Fed vs Starved tRNA LFCs")+
    ggsave(filename = paste(iso, "_full3prime_combinedonly_LFCs.png", sep = "_"), bg = "white", width = 15, height = 7, dpi = 700)
  
  plot_list[[iso]] <- temp_plot
  
  print(temp_plot)
}

#################################################################################################################################
# Visulisation of the DESeq2 data 
#################################################################################################################################

# P-VALUES HISTOGRAM SANITY CHECK 

png("pvalues_histogram.png")

hist(FedvsStarved$pvalue)

dev.off()

# FDR-VALUES HISTOGRAM SANITY CHECK

png("fdrvalues_histogram.png")

hist(FedvsStarved$padj)

dev.off()

#################################################################################################################################

# MA-PLOT POST DE

png("MA_plot_FedvsStarved.png")

plotMA(FedvsStarved, alpha = 0.05,
       main = "Fed vs Starved in BJ5TA Fibroblasts")

dev.off()

#################################################################################################################################

# Volvano Plot

library(tidyverse)
library(ggplot2)

df <- as.data.frame(FedvsStarved_table)

df %>%
  mutate(DE = factor(case_when(FDR < 0.05 & log2FC > 1 ~ "Up",
                               FDR < 0.05 & log2FC < -1 ~ "Down",
                               FDR >= 0.05 ~ "NS"), 
                     ordered = T, 
                     levels = c("Up", "Down", "NS"), 
                     labels = c("Up", "Down", "Not Sig"))) %>%
  arrange(FDR) -> new_df

new_df <- na.omit(new_df)

png("FedvsStarved_volcanoplot.png")

ggplot(data = new_df, aes(x = log2FC, y = -log10(FDR), colour = DE))+
  geom_point(size = 1, alpha = 0.5)+
  geom_vline(xintercept = -1.5, linetype = "dashed")+
  geom_vline(xintercept = 1.5, linetype = "dashed")+
  scale_colour_manual(values = c("Up" = "red", "Down" = "blue", "Not Sig" = "grey"))+
  xlim(-8, 8)+
  ylim(0, 310)+
  scale_x_continuous(breaks = seq(-8, 8, by = 2))+
  scale_y_continuous(breaks = seq(0, 300, by = 50))+
  geom_hline(yintercept = 1.30103, linetype = "dashed")+
  ylab("-log10(FDR)")+
  xlab("log2FC")+
  ggtitle("Fed vs Starved in BJ5TA Fibroblasts")+
  theme_bw()+
  theme(axis.title = element_text(size = 20),
        axis.text = element_text(size = 18),
        legend.title = element_blank(),
        legend.text = element_text(size = 20),
        plot.title = element_text(size = 20, hjust = 0.5),
        panel.grid = element_blank())

dev.off()

#################################################################################################################################

# Interactive StripChart with Glimma

library(Glimma)

group <- sample_information$Condition

de <- as.integer(FedvsStarved$padj <= 0.05 & !is.na(FedvsStarved$padj))

normCounts <- log2(counts(ddsObj))

glXYPlot(
  x = FedvsStarved$log2FoldChange,
  y = -log10(FedvsStarved$pvalue),
  xlab = "log2FC",
  ylab = "-log10(FDR)",
  side.xlab = "Condition",
  side.ylab = "Log(CPM)",
  main = "Fed vs Starved in BJ5TA Fibroblasts",
  counts = normCounts,
  groups = group,
  sample.cols = conditionCol,
  status = de,
  #anno = shrink_Day4_ConVsCre[, c("GeneID", "Symbol", "Description")],
  folder = "volcano")

#################################################################################################################################

# HEATMAPS

library(ComplexHeatmap)
library(circlize)

# get the top 200 genes
sigGenes <- as.data.frame(FedvsStarved_table) %>% 
  top_n(200, wt=-FDR) %>% 
  pull("Gene")

# filter the data for the top 200 by padj in the LRT test
plotDat <- varianceStabilizingTransformation(ddsObj)[sigGenes,] %>% 
  assay()
z.mat <- t(scale(t(plotDat), center=TRUE, scale=TRUE))

# colour palette
myPalette <- c("red3", "ivory", "blue3")
myRamp = colorRamp2(c(-2, 0, 2), myPalette)

Heatmap(z.mat, name = "z-score",
        col = myRamp,            
        show_row_names = FALSE,
        cluster_columns = FALSE)

# cluster the data and split the tree
hcDat <- hclust(dist(z.mat))
cutGroups <- cutree(hcDat, h=4)

ha1 = HeatmapAnnotation(df = colData(ddsObj)[,c("Condition", "batch")])

png("heatmap.png")

Heatmap(z.mat, name = "z-score",
        col = myRamp,            
        show_row_name = FALSE,
        cluster_columns = FALSE,
        split=cutGroups,
        rect_gp = gpar(col = "darkgrey", lwd=0.5),
        top_annotation = ha1)

dev.off()

#################################################################################################################################
# counts for collapsed anticodons
#################################################################################################################################
# need to take the originals counts table, and then do the following steps 
# like the barplot input, we need to get rid of any small RNAs
# seperate the tRNA name strings 
# delete the gene and loci columns 
# sum the counts by anticodon
# rerun the DESeq2 on the new collapsed counts and then plot  

anticodon_counts <- read_tsv("/mnt/data/BMOHAMED/TRAX_nov2019/tRAX-master/nomultimap_FedvsStarved/counts_analysis/tRNAsonly/raw_counts_tRNAsonly.txt", comment = "#")

newcolnames <- c("Gene", "Fed1", "Fed2", "Fed5", "Starved1", "Starved2", "Starved5")
colnames(anticodon_counts) <- newcolnames

# place all the small RNAs into a seperate table for now ... all tRNAs begin with "tRNA-"

smallRNAcounts <- filter(anticodon_counts, !grepl("tRNA-", Gene)) 

# place all of the tRNAs into a seperate table 

anticodon_counts <- filter(anticodon_counts, grepl("tRNA-", Gene))

# split Gene column in seperate columns to get the anticodons 

anticodon_counts <- separate(anticodon_counts, Gene, sep = "-", into = c("RNA", "Isodecoder", "Anticodon", "Loci"))

# remove the Gene and tRNA rows because they won't be needed in this analysis 

anticodon_counts <- select(anticodon_counts, -c(RNA, Loci))

# sum the counts by anticodon

anticodon_counts <- aggregate(as.matrix(anticodon_counts[,3:8]), as.list(anticodon_counts[,1:2]), FUN = sum)

# make a new column called label with the AA and anticodon concatenated (to be used later in the barplots)

anticodon_counts <- mutate(anticodon_counts, Label = paste(Isodecoder, Anticodon, sep = "-"), Gene = paste(Isodecoder, Anticodon, sep = "-"))

# make a seperate table with just the Gene column and the counts for each condition

anticodon_counts_DESeq2input <- select(anticodon_counts, -c(Isodecoder, Anticodon, Label))

# merge the anticodon counts table and the small RNAs table ... making the "Gene" the first column in the table 

anticodon_counts_DESeq2input <- union(anticodon_counts_DESeq2input, smallRNAcounts) %>%
  select(Gene, everything()) # note that there are no small RNAs in this analysis

# save this counts table and the same above analysis will be done in a seperate R document 

write_tsv(anticodon_counts_DESeq2input, "/mnt/data/BMOHAMED/TRAX_nov2019/tRAX-master/resecregen/counts_analysis/collapsed_anticodons/anticodon_counts_DESeq2input.txt")


#################################################################################################################################
# generate normalized counts & panel to LFCs
#################################################################################################################################

#################################################################################################################################
# Barplot of AT ending tRNAs 
#################################################################################################################################
# Here, I'm filtering all anticodons that begin with either an A or T 
# remember that we are looking at tRNAs and so the anticodon is in reverse complement (i.e. first position)

at_ending <- barplot_input %>%
  filter(str_detect(Anticodon, "^A|^T"))

a_ending <- barplot_input %>%
  filter(str_detect(Anticodon, "^A"))

t_ending <- barplot_input %>%
  filter(str_detect(Anticodon, "^T"))

plot_list <- list()

ggplot(data = at_ending, aes(x = Label, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(at_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(at_ending$FDR < 0.05, 6, 4))+
  xlab("Isodecoder")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("A/T at the anticodon wobble position - Fed vs Starved in BJ5TA Fibroblasts")+
  ggsave(filename = paste("AT_ending_tRNA_LFCs.png"), bg = "white", width = 15, height = 7, dpi = 700)

plot_list[[iso]] <- temp_plot


#################################################################################################################################
# Barplot of GC ending tRNAs 
#################################################################################################################################
# Here, I'm filtering all anticodons that begin with either a G or C 
# remember that we are looking at tRNAs and so the anticodon is in reverse complement (i.e. first position)

gc_ending <- barplot_input %>%
  filter(str_detect(Anticodon, "^G|^C"))

g_ending <- barplot_input %>%
  filter(str_detect(Anticodon, "^G"))

c_ending <- barplot_input %>%
  filter(str_detect(Anticodon, "^C"))

plot_list <- list()

ggplot(data = gc_ending, aes(x = Label, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(gc_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(gc_ending$FDR < 0.05, 6, 4))+
  xlab("Isodecoder")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("G/C at the anticodon wobble position - Fed vs Starved in BJ5TA Fibroblasts")+
  ggsave(filename = paste("GC_ending_tRNA_LFCs.png"), bg = "white", width = 15, height = 7, dpi = 700)

plot_list[[iso]] <- temp_plot


#################################################################################################################################
# plots of the normalized counts for each replicate 
#################################################################################################################################

# normalize the raw counts 

normalized_counts <- seq_data

barplot_input <- mutate(barplot_input, Gene = Label)

# delete the rows that are not present in the barplot_input

normalized_counts <- normalized_counts[(normalized_counts$Gene %in% barplot_input$Gene),]

# divide all the counts by their respective size factors 

normalized_counts <- mapply("/", normalized_counts[intersect(names(normalized_counts), names(sizeFactors))], sizeFactors[intersect(names(normalized_counts), names(sizeFactors))])

# combine the barplot table to get the gene names again 

normalized_counts <- cbind(barplot_input, normalized_counts)

# removing columns that are not needed 

normalized_counts <- select(normalized_counts, -c(5,7,8,11))

# calculate the mean of the normalized counts and add a new column for each 

normalized_counts <- mutate(normalized_counts, Fed_Mean = rowMeans(cbind(normalized_counts$Fed1, normalized_counts$Fed2, normalized_counts$Fed5)))

normalized_counts <- mutate(normalized_counts, Starved_Mean = rowMeans(cbind(normalized_counts$Starved1, normalized_counts$Starved2, normalized_counts$Starved5)))

# plot scatter graph for each anticodon with the mean as a line going through the scatter 

library(ggplot2)
library(grid)
library(gridExtra)
library(reshape2)
library(tidyr)
library(cowplot)

df3 <- normalized_counts

df3_melt <- tidyr::gather(df3, variable, value, Fed1:Starved5)

df3_melt <- mutate(df3_melt, Condition = case_when(variable == "Fed1" ~ "Fed",
                                                   variable == "Fed2" ~ "Fed",
                                                   variable == "Fed5" ~ "Fed",
                                                   variable == "Starved1" ~ "Starved",
                                                   variable == "Starved2" ~ "Starved",
                                                   variable == "Starved5" ~ "Starved"))


for (iso in c("Ala", "Arg", "Asn", "Asp", "Cys", "Gln", "Glu", "Gly", "His", "Ile", "Leu", "Lys", "Met", "Phe", "Pro", "SeC", "Ser", "Thr", "Trp", "Tyr", "Val", "iMet")) {
  
  df4 <- subset(df3_melt, Isodecoder == iso) 
  
  normalized_counts_graphs <- ggplot(data = df4, aes(x = Label, y = value, colour = Condition, group = Condition))+
    geom_point(position = position_dodge(width = 0.2))+
    #facet_wrap( ~ Condition)+
    #stat_summary(fun.y = mean, aes(group = 1), geom = "point", colour = "black", shape = 4, size = 5)+
    #ggrepel::geom_text_repel(aes(label = variable), colour = "black", size = 2.5, segment.colour = "grey")+
    xlab("Isodecoder")+
    ylab("Normalized Counts")+
    theme_bw()+
    theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
    ggtitle(iso, "Fed vs Starved - Normalized Counts")
  
  df2 <- subset(barplot_input, Isodecoder == iso) 
  
  tRNA_barplot <- ggplot(data = df2, aes(x = Label, y = log2FC, fill = Anticodon, label = ifelse(FDR < 0.05, "*", "ns")))+
    geom_bar(stat = "identity")+
    geom_text(vjust = 0, nudge_y = ifelse(df2$log2FC < 0, -1, 1), size = ifelse(df2$FDR < 0.05, 6, 4))+
    xlab("Isodecoder")+
    ylab(expression(paste(Log[2],"FC",sep="")))+
    ylim(c(-5, 5))+
    theme_bw()+
    theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
    ggtitle(iso, "Fed vs Starved - tRNA LFCs")
  
  grid_panel <- plot_grid(normalized_counts_graphs, tRNA_barplot, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)
  
  save_plot(filename = paste(iso, "panel_tRNAsonly.png", sep = "_"), grid_panel, nrow = 2)    
}

# plots with only the mean normalized counts 

df3_mean <- normalized_counts

df3_melt_mean <- tidyr::gather(df3_mean, variable, value, Fed_Mean:Starved_Mean)

df3_melt_mean <- mutate(df3_melt_mean, Condition = case_when(variable == "Fed_Mean" ~ "Fed - Mean of the normalized counts",
                                                             variable == "Starved_Mean" ~ "Starved - Mean of the normalized counts"))


for (iso in c("Ala", "Arg", "Asn", "Asp", "Cys", "Gln", "Glu", "Gly", "His", "Ile", "Leu", "Lys", "Met", "Phe", "Pro", "SeC", "Ser", "Thr", "Trp", "Tyr", "Val", "iMet")) {
  
  df4_mean <- subset(df3_melt_mean, Isodecoder == iso) 
  
  normalized_counts_graphs_mean <- ggplot(data = df4_mean, aes(x = Label, y = value, colour = Condition, group = Condition))+
    geom_point(position = position_dodge(width = 0.2))+
    #facet_wrap( ~ Condition)+
    #stat_summary(fun.y = mean, aes(group = 1), geom = "point", colour = "black", shape = 4, size = 5)+
    #ggrepel::geom_text_repel(aes(label = variable), colour = "black", size = 2.5, segment.colour = "grey")+
    xlab("Isodecoder")+
    ylab("Normalized Counts")+
    theme_bw()+
    theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
    ggtitle(iso, "Fed vs Starved - Mean Normalized Counts")
  
  df2 <- subset(barplot_input, Isodecoder == iso) 
  
  tRNA_barplot_mean <- ggplot(data = df2, aes(x = Label, y = log2FC, fill = Anticodon, label = ifelse(FDR < 0.05, "*", "ns")))+
    geom_bar(stat = "identity")+
    geom_text(vjust = 0, nudge_y = ifelse(df2$log2FC < 0, -1, 1), size = ifelse(df2$FDR < 0.05, 6, 4))+
    xlab("Isodecoder")+
    ylab(expression(paste(Log[2],"FC",sep="")))+
    ylim(c(-5, 5))+
    theme_bw()+
    theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
    ggtitle(iso, "Fed vs Starved - tRNA LFCs")
  
  grid_panel <- plot_grid(normalized_counts_graphs_mean, tRNA_barplot_mean, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)
  
  save_plot(filename = paste(iso, "panel_tRNAsonly_mean.png", sep = "_"), grid_panel, nrow = 2)    
}

#################################################################################################################################
# plots of the normalized counts for each replicate ... individual plots 
#################################################################################################################################

for (iso in c("Ala", "Arg", "Asn", "Asp", "Cys", "Gln", "Glu", "Gly", "His", "Ile", "Leu", "Lys", "Met", "Phe", "Pro", "SeC", "Ser", "Thr", "Trp", "Tyr", "Val", "iMet")) {
  
  df4 <- subset(df3_melt, Isodecoder == iso) 
  
  normalized_counts_graphs <- ggplot(data = df4, aes(x = Label, y = value, colour = Condition, group = Condition))+
    geom_point(position = position_dodge(width = 0.2))+
    #facet_wrap( ~ Condition)+
    #stat_summary(fun.y = mean, aes(group = 1), geom = "point", colour = "black", shape = 4, size = 5)+
    #ggrepel::geom_text_repel(aes(label = variable), colour = "black", size = 2.5, segment.colour = "grey")+
    xlab("Isodecoder")+
    ylab("Normalized Counts")+
    theme_bw()+
    theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
    ggtitle(iso, "Fed vs Starved - Normalized Counts")+
    ggsave(filename = paste(iso, "_full3primetRNAsonly_normalizedcounts.png", sep = "_"), bg = "white", width = 15, height = 7, dpi = 700)
}

#################################################################################################################################
# plots of the normalized counts AND LFC depending on wobble position 
#################################################################################################################################

# GC at the wobble position 

gc_ending_figure <- normalized_counts %>%
  filter(str_detect(Anticodon, "^G|^C"))

df_gc_wobble_melt <- tidyr::gather(gc_ending_figure, variable, value, Fed1:Starved5) 

df_gc_wobble_melt <- mutate(df_gc_wobble_melt, Condition = case_when(variable == "Fed1" ~ "Fed",
                                                                     variable == "Fed2" ~ "Fed",
                                                                     variable == "Fed5" ~ "Fed",
                                                                     variable == "Starved1" ~ "Starved",
                                                                     variable == "Starved2" ~ "Starved",
                                                                     variable == "Starved5" ~ "Starved"))
#variable == "Fed_Mean" ~ "Fed_Mean",
#variable == "Starved_Mean" ~ "Starved_Mean"))


normalized_counts_wobble_graphs_gc <- ggplot(data = df_gc_wobble_melt, aes(x = Label, y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  xlab("Isodecoder")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("G/C at the anticodon wobble position", "Fed vs Starved - Normalized Counts")

GC_at_wobble <- ggplot(data = gc_ending, aes(x = Label, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(gc_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(gc_ending$FDR < 0.05, 6, 4))+
  xlab("Isodecoder")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("G/C at the anticodon wobble position", "Fed vs Starved - tRNA LFCs")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_gc, GC_at_wobble, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "GC_ending_tRNAs_panel.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)




# AT at the wobble position  

at_ending_figure <- normalized_counts %>%
  filter(str_detect(Anticodon, "^A|^T"))

df_at_wobble_melt <- tidyr::gather(at_ending_figure, variable, value, Fed1:Starved5) 

df_at_wobble_melt <- mutate(df_at_wobble_melt, Condition = case_when(variable == "Fed1" ~ "Fed",
                                                                     variable == "Fed2" ~ "Fed",
                                                                     variable == "Fed5" ~ "Fed",
                                                                     variable == "Starved1" ~ "Starved",
                                                                     variable == "Starved2" ~ "Starved",
                                                                     variable == "Starved5" ~ "Starved"))


normalized_counts_wobble_graphs_at <- ggplot(data = df_at_wobble_melt, aes(x = Label, y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  xlab("Isodecoder")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("A/T at the anticodon wobble position", "Fed vs Starved - Normalized Counts")

AT_at_wobble <- ggplot(data = at_ending, aes(x = Label, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(at_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(at_ending$FDR < 0.05, 6, 4))+
  xlab("Isodecoder")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("A/T at the anticodon wobble position", "Fed vs Starved")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_at, AT_at_wobble, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "AT_ending_tRNAs_panel.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)


#################################################################################################################################
# plots of the normalized counts AND LFC depending on wobble position ... only showing the mean normalized count
#################################################################################################################################


# GC at the wobble position 

gc_ending_figure_mean <- normalized_counts %>%
  filter(str_detect(Anticodon, "^G|^C"))

df_gc_wobble_melt_mean <- tidyr::gather(gc_ending_figure_mean, variable, value, Fed_Mean:Starved_Mean) 

df_gc_wobble_melt_mean <- mutate(df_gc_wobble_melt_mean, Condition = case_when(variable == "Fed_Mean" ~ "Fed - Mean of the normalized counts",
                                                                               variable == "Starved_Mean" ~ "Starved - Mean of the normalized counts"))


normalized_counts_wobble_graphs_gc_mean <- ggplot(data = df_gc_wobble_melt_mean, aes(x = Label, y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  xlab("Isodecoder")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("G/C at the anticodon wobble position", "Fed vs Starved - Normalized Counts")

GC_at_wobble_mean <- ggplot(data = gc_ending, aes(x = Label, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(gc_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(gc_ending$FDR < 0.05, 6, 4))+
  xlab("Isodecoder")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("G/C at the anticodon wobble position", "Fed vs Starved - tRNA LFCs")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_gc_mean, GC_at_wobble_mean, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "GC_ending_tRNAs_panel_mean.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)




# AT at the wobble position  

at_ending_figure_mean <- normalized_counts %>%
  filter(str_detect(Anticodon, "^A|^T"))

df_at_wobble_melt_mean <- tidyr::gather(at_ending_figure_mean, variable, value, Fed_Mean:Starved_Mean) 

df_at_wobble_melt_mean <- mutate(df_at_wobble_melt_mean, Condition = case_when(variable == "Fed_Mean" ~ "Fed - Mean of the normalized counts",
                                                                               variable == "Starved_Mean" ~ "Starved - Mean of the normalized counts"))


normalized_counts_wobble_graphs_at_mean <- ggplot(data = df_at_wobble_melt_mean, aes(x = Label, y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  xlab("Isodecoder")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("A/T at the anticodon wobble position", "Fed vs Starved - Normalized Counts")

AT_at_wobble_mean <- ggplot(data = at_ending, aes(x = Label, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(at_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(at_ending$FDR < 0.05, 6, 4))+
  xlab("Isodecoder")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("A/T at the anticodon wobble position", "Fed vs Starved - tRNA LFCs")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_at_mean, AT_at_wobble_mean, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "AT_ending_tRNAs_panel_mean.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)


#################################################################################################################################
# seperating counts into individual Wobble position nucleotides 
#################################################################################################################################

# A at the wobble position  

a_ending_figure <- normalized_counts %>%
  filter(str_detect(Anticodon, "^A"))

df_a_wobble_melt <- tidyr::gather(a_ending_figure, variable, value, Fed1:Starved5) 

df_a_wobble_melt <- mutate(df_a_wobble_melt, Condition = case_when(variable == "Fed1" ~ "Fed",
                                                                   variable == "Fed2" ~ "Fed",
                                                                   variable == "Fed5" ~ "Fed",
                                                                   variable == "Starved1" ~ "Starved",
                                                                   variable == "Starved2" ~ "Starved",
                                                                   variable == "Starved5" ~ "Starved"))


normalized_counts_wobble_graphs_a <- ggplot(data = df_a_wobble_melt, aes(x = Label, y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  xlab("Isodecoder")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("A at the anticodon wobble position", "Fed vs Starved - Normalized Counts")

A_at_wobble <- ggplot(data = a_ending, aes(x = Label, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(a_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(a_ending$FDR < 0.05, 6, 4))+
  xlab("Isodecoder")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("A at the anticodon wobble position", "Fed vs Starved - Log2FC")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_a, A_at_wobble, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "A_ending_tRNAs_panel.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)

#################################################################################################################################

# T at the wobble position  

t_ending_figure <- normalized_counts %>%
  filter(str_detect(Anticodon, "^T"))

df_t_wobble_melt <- tidyr::gather(t_ending_figure, variable, value, Fed1:Starved5) 

df_t_wobble_melt <- mutate(df_t_wobble_melt, Condition = case_when(variable == "Fed1" ~ "Fed",
                                                                   variable == "Fed2" ~ "Fed",
                                                                   variable == "Fed5" ~ "Fed",
                                                                   variable == "Starved1" ~ "Starved",
                                                                   variable == "Starved2" ~ "Starved",
                                                                   variable == "Starved5" ~ "Starved"))


normalized_counts_wobble_graphs_t <- ggplot(data = df_t_wobble_melt, aes(x = Label, y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  xlab("Isodecoder")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("T at the anticodon wobble position", "Fed vs Starved - Normalized Counts")

T_at_wobble <- ggplot(data = t_ending, aes(x = Label, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(t_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(t_ending$FDR < 0.05, 6, 4))+
  xlab("Isodecoder")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("T at the anticodon wobble position", "Fed vs Starved - Log2FC")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_t, T_at_wobble, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "T_ending_tRNAs_panel.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)

#################################################################################################################################

# G at the wobble position  

g_ending_figure <- normalized_counts %>%
  filter(str_detect(Anticodon, "^G"))

df_g_wobble_melt <- tidyr::gather(g_ending_figure, variable, value, Fed1:Starved5) 

df_g_wobble_melt <- mutate(df_g_wobble_melt, Condition = case_when(variable == "Fed1" ~ "Fed",
                                                                   variable == "Fed2" ~ "Fed",
                                                                   variable == "Fed5" ~ "Fed",
                                                                   variable == "Starved1" ~ "Starved",
                                                                   variable == "Starved2" ~ "Starved",
                                                                   variable == "Starved5" ~ "Starved"))


normalized_counts_wobble_graphs_g <- ggplot(data = df_g_wobble_melt, aes(x = Label, y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  xlab("Isodecoder")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("G at the anticodon wobble position", "Fed vs Starved - Normalized Counts")

G_at_wobble <- ggplot(data = g_ending, aes(x = Label, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(g_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(g_ending$FDR < 0.05, 6, 4))+
  xlab("Isodecoder")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("G at the anticodon wobble position", "Fed vs Starved - Log2FC")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_g, G_at_wobble, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "G_ending_tRNAs_panel.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)

#################################################################################################################################

# C at the wobble position  

c_ending_figure <- normalized_counts %>%
  filter(str_detect(Anticodon, "^C"))

df_c_wobble_melt <- tidyr::gather(c_ending_figure, variable, value, Fed1:Starved5) 

df_c_wobble_melt <- mutate(df_c_wobble_melt, Condition = case_when(variable == "Fed1" ~ "Fed",
                                                                   variable == "Fed2" ~ "Fed",
                                                                   variable == "Fed5" ~ "Fed",
                                                                   variable == "Starved1" ~ "Starved",
                                                                   variable == "Starved2" ~ "Starved",
                                                                   variable == "Starved5" ~ "Starved"))


normalized_counts_wobble_graphs_c <- ggplot(data = df_c_wobble_melt, aes(x = Label, y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  xlab("Isodecoder")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("C at the anticodon wobble position", "Fed vs Starved - Normalized Counts")

C_at_wobble <- ggplot(data = c_ending, aes(x = Label, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(c_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(c_ending$FDR < 0.05, 6, 4))+
  xlab("Isodecoder")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("C at the anticodon wobble position", "Fed vs Starved - Log2FC")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_c, C_at_wobble, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "C_ending_tRNAs_panel.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)

#################################################################################################################################
# Place all A/T/G/C ending wobble positions into one panel
#################################################################################################################################

grid_panel_ATCG_wobble <- plot_grid(A_at_wobble, T_at_wobble, G_at_wobble, C_at_wobble, ncol = 2, labels = c("A", "B", "C", "D"))

save_plot(filename = "ATGC_ending_tRNAs_panel.png", grid_panel_ATCG_wobble, nrow = 2, base_width = 15, base_height = 7)


#################################################################################################################################
# A/T/C/G normalized counts individual graphs 
#################################################################################################################################

ggsave(plot = normalized_counts_wobble_graphs_a, filename = "normalized_counts_wobble_graphs_a.png", bg = "white", width = 15, height = 7, dpi = 700)

ggsave(plot = normalized_counts_wobble_graphs_t, filename = "normalized_counts_wobble_graphs_t.png", bg = "white", width = 15, height = 7, dpi = 700)

ggsave(plot = normalized_counts_wobble_graphs_c, filename = "normalized_counts_wobble_graphs_c.png", bg = "white", width = 15, height = 7, dpi = 700)

ggsave(plot = normalized_counts_wobble_graphs_g, filename = "normalized_counts_wobble_graphs_g.png", bg = "white", width = 15, height = 7, dpi = 700)

#################################################################################################################################
# log scale graphs 
#################################################################################################################################

normalized_counts_wobble_graphs_c_log <- normalized_counts_wobble_graphs_c +
  scale_y_continuous(trans = "log2")+
  ggtitle("C at the anticodon wobble position", "Fed vs Starved - Normalized Counts (Log2 Scale)")

ggsave(plot = normalized_counts_wobble_graphs_c_log, filename = "normalized_counts_wobble_graphs_c_log.png", bg = "white", width = 15, height = 7, dpi = 700)

normalized_counts_wobble_graphs_a_log <- normalized_counts_wobble_graphs_a +
  ggtitle("A at the anticodon wobble position", "Fed vs Starved - Normalized Counts (Log2 Scale)")+
  scale_y_continuous(trans = "log2")

ggsave(plot = normalized_counts_wobble_graphs_a_log, filename = "normalized_counts_wobble_graphs_a_log.png", bg = "white", width = 15, height = 7, dpi = 700)

normalized_counts_wobble_graphs_t_log <- normalized_counts_wobble_graphs_t +
  ggtitle("T at the anticodon wobble position", "Fed vs Starved - Normalized Counts (Log2 Scale)")+
  scale_y_continuous(trans = "log2")

ggsave(plot = normalized_counts_wobble_graphs_t_log, filename = "normalized_counts_wobble_graphs_t_log.png", bg = "white", width = 15, height = 7, dpi = 700)

normalized_counts_wobble_graphs_g_log <- normalized_counts_wobble_graphs_g +
  ggtitle("G at the anticodon wobble position", "Fed vs Starved - Normalized Counts (Log2 Scale)")+
  scale_y_continuous(trans = "log2")

ggsave(plot = normalized_counts_wobble_graphs_g_log, filename = "normalized_counts_wobble_graphs_g_log.png", bg = "white", width = 15, height = 7, dpi = 700)

#################################################################################################################################
# log2 scale graphs 
#################################################################################################################################

# A at the wobble position  

grid_panel <- plot_grid(normalized_counts_wobble_graphs_a_log, A_at_wobble, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "A_ending_tRNAs_panel_log.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)

#################################################################################################################################

# T at the wobble position  

grid_panel <- plot_grid(normalized_counts_wobble_graphs_t_log, T_at_wobble, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "T_ending_tRNAs_panel_log.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)

#################################################################################################################################

# G at the wobble position  

grid_panel <- plot_grid(normalized_counts_wobble_graphs_g_log, G_at_wobble, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "G_ending_tRNAs_panel_log.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)

#################################################################################################################################

# C at the wobble position  

grid_panel <- plot_grid(normalized_counts_wobble_graphs_c_log, C_at_wobble, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "C_ending_tRNAs_panel_log.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)

#################################################################################################################################
# log2 scale plots of the normalized counts for each replicate ... individual plots 

for (iso in c("Ala", "Arg", "Asn", "Asp", "Cys", "Gln", "Glu", "Gly", "His", "Ile", "Leu", "Lys", "Met", "Phe", "Pro", "SeC", "Ser", "Thr", "Trp", "Tyr", "Val", "iMet")) {
  
  df4 <- subset(df3_melt, Isodecoder == iso) 
  
  normalized_counts_graphs_logscale <- ggplot(data = df4, aes(x = Label, y = value, colour = Condition, group = Condition))+
    geom_point(position = position_dodge(width = 0.2))+
    xlab("Isodecoder")+
    ylab("Normalized Counts")+
    scale_y_continuous(trans = "log2")+
    theme_bw()+
    theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
    ggtitle(iso, "Fed vs Starved - Normalized Counts (Log2 Scale)")+
    ggsave(filename = paste(iso, "_full3primetRNAsonly_normalizedcounts_log.png", sep = "_"), bg = "white", width = 15, height = 7, dpi = 700)
}

#################################################################################################################################
# log2 scale panel plots

for (iso in c("Ala", "Arg", "Asn", "Asp", "Cys", "Gln", "Glu", "Gly", "His", "Ile", "Leu", "Lys", "Met", "Phe", "Pro", "SeC", "Ser", "Thr", "Trp", "Tyr", "Val", "iMet")) {
  
  df4 <- subset(df3_melt, Isodecoder == iso) 
  
  normalized_counts_graphs_logscale <- ggplot(data = df4, aes(x = Label, y = value, colour = Condition, group = Condition))+
    geom_point(position = position_dodge(width = 0.2))+
    scale_y_continuous(trans = "log2")+
    xlab("Isodecoder")+
    ylab("Normalized Counts")+
    theme_bw()+
    theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
    ggtitle(iso, "Fed vs Starved - Normalized Counts (Log2 Scale)")
  
  df2 <- subset(barplot_input, Isodecoder == iso) 
  
  tRNA_barplot <- ggplot(data = df2, aes(x = Label, y = log2FC, fill = Anticodon, label = ifelse(FDR < 0.05, "*", "ns")))+
    geom_bar(stat = "identity")+
    geom_text(vjust = 0, nudge_y = ifelse(df2$log2FC < 0, -1, 1), size = ifelse(df2$FDR < 0.05, 6, 4))+
    xlab("Isodecoder")+
    ylab(expression(paste(Log[2],"FC",sep="")))+
    ylim(c(-5, 5))+
    theme_bw()+
    theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
    ggtitle(iso, "Fed vs Starved - tRNA LFCs")
  
  grid_panel <- plot_grid(normalized_counts_graphs_logscale, tRNA_barplot, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)
  
  save_plot(filename = paste(iso, "panel_tRNAsonly_logscale.png", sep = "_"), grid_panel, nrow = 2, bg = "white", base_width = 15, base_height = 7, dpi = 500)    
}

#################################################################################################################################
# Rank the log scale graphs
#################################################################################################################################

normalized_counts_wobble_graphs_c_log_ordered <- ggplot(data = df_c_wobble_melt, aes(x = reorder(Label, -value), y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  scale_y_continuous(trans = "log2") +
  xlab("Isodecoder")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("C at the anticodon wobble position", "Fed vs Starved - Normalized Counts (Log2 Scale)")

ggsave(plot = normalized_counts_wobble_graphs_c_log_ordered, filename = "normalized_counts_wobble_graphs_c_log_ordered.png", bg = "white", width = 15, height = 7, dpi = 700)

normalized_counts_wobble_graphs_a_log_ordered <- ggplot(data = df_a_wobble_melt, aes(x = reorder(Label, -value), y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  scale_y_continuous(trans = "log2") +
  xlab("Isodecoder")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("A at the anticodon wobble position", "Fed vs Starved - Normalized Counts (Log2 Scale)")

ggsave(plot = normalized_counts_wobble_graphs_a_log_ordered, filename = "normalized_counts_wobble_graphs_a_log_ordered.png", bg = "white", width = 15, height = 7, dpi = 700)

normalized_counts_wobble_graphs_t_log_ordered <- ggplot(data = df_t_wobble_melt, aes(x = reorder(Label, -value), y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  scale_y_continuous(trans = "log2") +
  xlab("Isodecoder")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("T at the anticodon wobble position", "Fed vs Starved - Normalized Counts (Log2 Scale)")

ggsave(plot = normalized_counts_wobble_graphs_t_log_ordered, filename = "normalized_counts_wobble_graphs_t_log_ordered.png", bg = "white", width = 15, height = 7, dpi = 700)

normalized_counts_wobble_graphs_g_log_ordered <- ggplot(data = df_g_wobble_melt, aes(x = reorder(Label, -value), y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  scale_y_continuous(trans = "log2") +
  xlab("Isodecoder")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("G at the anticodon wobble position", "Fed vs Starved - Normalized Counts (Log2 Scale)")

ggsave(plot = normalized_counts_wobble_graphs_g_log_ordered, filename = "normalized_counts_wobble_graphs_g_log_ordered.png", bg = "white", width = 15, height = 7, dpi = 700)

#################################################################################################################################

# A at the wobble position 

normalized_counts_wobble_graphs_a_log_ordered_log2FC <- ggplot(data = df_a_wobble_melt, aes(x = reorder(Label, -log2FC), y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  scale_y_continuous(trans = "log2") +
  xlab("Isodecoder")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 14, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 14))+
  ggtitle("A at the anticodon wobble position", "Fed vs Starved - Normalized Counts (Log2 Scale)")

A_at_wobble_ordered <- ggplot(data = a_ending, aes(x = reorder(Label, -log2FC), y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(a_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(a_ending$FDR < 0.05, 6, 4))+
  xlab("Isodecoder")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 14, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 14))+
  ggtitle("A at the anticodon wobble position", "Fed vs Starved")


grid_panel <- plot_grid(normalized_counts_wobble_graphs_a_log_ordered_log2FC, A_at_wobble_ordered, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "A_ending_tRNAs_panel_log_ordered.png", grid_panel, nrow = 2, base_width = 20, base_height = 8)

#################################################################################################################################

# T at the wobble position  

normalized_counts_wobble_graphs_t_log_ordered_log2FC <- ggplot(data = df_t_wobble_melt, aes(x = reorder(Label, -log2FC), y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  scale_y_continuous(trans = "log2") +
  xlab("Isodecoder")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 14, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 14))+
  ggtitle("T at the anticodon wobble position", "Fed vs Starved - Normalized Counts (Log2 Scale)")

T_at_wobble_ordered <- ggplot(data = t_ending, aes(x = reorder(Label, -log2FC), y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(t_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(t_ending$FDR < 0.05, 6, 4))+
  xlab("Isodecoder")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 14, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 14))+
  ggtitle("T at the anticodon wobble position", "Fed vs Starved")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_t_log_ordered_log2FC, T_at_wobble_ordered, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "T_ending_tRNAs_panel_log_ordered.png", grid_panel, nrow = 2, base_width = 20, base_height = 8)

#################################################################################################################################

# G at the wobble position  

normalized_counts_wobble_graphs_g_log_ordered_log2FC <- ggplot(data = df_g_wobble_melt, aes(x = reorder(Label, -log2FC), y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  scale_y_continuous(trans = "log2") +
  xlab("Isodecoder")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 14, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 14))+
  ggtitle("G at the anticodon wobble position", "Fed vs Starved - Normalized Counts (Log2 Scale)")

G_at_wobble_ordered <- ggplot(data = g_ending, aes(x = reorder(Label, -log2FC), y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(g_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(g_ending$FDR < 0.05, 6, 4))+
  xlab("Isodecoder")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 14, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 14))+
  ggtitle("G at the anticodon wobble position", "Fed vs Starved")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_g_log_ordered_log2FC, G_at_wobble_ordered, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "G_ending_tRNAs_panel_log_ordered.png", grid_panel, nrow = 2, base_width = 20, base_height = 8)

#################################################################################################################################

# C at the wobble position  

normalized_counts_wobble_graphs_c_log_ordered_log2FC <- ggplot(data = df_c_wobble_melt, aes(x = reorder(Label, -log2FC), y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  scale_y_continuous(trans = "log2") +
  xlab("Isodecoder")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 14, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 14))+
  ggtitle("C at the anticodon wobble position", "Fed vs Starved - Normalized Counts (Log2 Scale)")

C_at_wobble_ordered <- ggplot(data = c_ending, aes(x = reorder(Label, -log2FC), y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(c_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(c_ending$FDR < 0.05, 6, 4))+
  xlab("Isodecoder")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  ylim(c(-4, 5))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 14, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 14))+
  ggtitle("C at the anticodon wobble position", "Fed vs Starved")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_c_log_ordered_log2FC, C_at_wobble_ordered, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "C_ending_tRNAs_panel_log_ordered.png", grid_panel, nrow = 2, base_width = 20, base_height = 8)

#################################################################################################################################
# log2 scale panel plots ordered by log2FC

for (iso in c("Ala", "Arg", "Asn", "Asp", "Cys", "Gln", "Glu", "Gly", "His", "Ile", "Leu", "Lys", "Met", "Phe", "Pro", "SeC", "Ser", "Thr", "Trp", "Tyr", "Val", "iMet")) {
  
  df4 <- subset(df3_melt, Isodecoder == iso) 
  
  normalized_counts_graphs_logscale_orderedLog2FC <- ggplot(data = df4, aes(x = reorder(Label, -log2FC), y = value, colour = Condition, group = Condition))+
    geom_point(position = position_dodge(width = 0.2))+
    scale_y_continuous(trans = "log2")+
    xlab("Isodecoder")+
    ylab("Normalized Counts")+ 
    theme_bw()+
    theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
    ggtitle(iso, "Fed vs Starved - Normalized Counts (Log2 Scale)")
  
  df2 <- subset(barplot_input, Isodecoder == iso) 
  
  tRNA_barplot_orderedLog2FC <- ggplot(data = df2, aes(x = reorder(Label, -log2FC), y = log2FC, fill = Anticodon, label = ifelse(FDR < 0.05, "*", "ns")))+
    geom_bar(stat = "identity")+
    geom_text(vjust = 0, nudge_y = ifelse(df2$log2FC < 0, -1, 1), size = ifelse(df2$FDR < 0.05, 6, 4))+
    xlab("Isodecoder")+
    ylab(expression(paste(Log[2],"FC",sep="")))+
    ylim(c(-5, 5))+
    theme_bw()+
    theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
    ggtitle(iso, "Fed vs Starved - tRNA LFCs")
  
  grid_panel <- plot_grid(normalized_counts_graphs_logscale_orderedLog2FC, tRNA_barplot_orderedLog2FC, ncol = 1, labels = c("A", "B"), rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)
  
  save_plot(filename = paste(iso, "panel_tRNAsonly_logscale_ordered.png", sep = "_"), grid_panel, nrow = 2, bg = "white", base_width = 15, base_height = 7, dpi = 500)    
}






