setwd("/mnt/data/BMOHAMED/TRAX_nov2019/tRAX-master/MDM2/counts_analysis/nocre1_2/collapsed_anticodon/")

library(DESeq2)
library(tidyverse)
library(dplyr)

#################################################################################################################################

# LOADING OUR DATASETS

sample_information <- read_tsv("samplefile_MDM2_males.txt")

seq_data <- read_tsv("/mnt/data/BMOHAMED/TRAX_nov2019/tRAX-master/MDM2/counts_analysis/nocre1_2/collapsed_anticodon/anticodon_counts_DESeq2input.txt")

seq_data <- seq_data[apply(seq_data[, -1], MARGIN = 1, function(x) all(x > 10)), ]

newcolnames <- c("Gene", "Null1", "Null2", "Null3", "Null4", "Null5", "Cre1", "Cre2", "Cre3")
colnames(seq_data) <- newcolnames

countdata <- seq_data %>%
  column_to_rownames("Gene") %>%
  as.matrix()

# Loaded the sample file containing (1) Bam file name (2) Sample name and (3) Condition
# Loaded the featureCounts.txt file generated by the counting step 
# comment = "#" comments out everything in the document accept the table itself 

#################################################################################################################################

# QUALITY ASSESMENT

# We can plot how many reads we have for each sample
# Although normalisation can account for imbalance, extreme differences may be indicitive of underlying problems 

library_sizes <- colSums(countdata)

conditionCol <- match(sample_information$Condition, c("Null", "Cre"))

png("library_sizes_barplot.png")

barplot(library_sizes,
        names=names(library_sizes),
        las=1,
        main = "Barplot of Library sizes",
        col = conditionCol)
#abline(h=15e7, lty=2)

dev.off()

# In the above code, we're taking the sum of all columns and assigning it the variable library_sizes
# We then use the R barplot function to plot a barplot of all the library sizes in each condition 
# names tells us what to call each bar
# las details the orientation of the x-axis label ... 1 = horrizontal 2= verticle
# main is the title of the plot 
# abline is the dotted line ... h = y-intercept lty= how dotted the line will be with 1=solid line 
# 14million was picked and is an arbitary number 


#################################################################################################################################

# COUNT DISTRIBUTION BOXPLOTS 

# Count data is not normally distributed and so to examine raw counts, its best to log transfrom them (log2)
# Count data may contain 0's and so to avoid log2(0), we add 1 to every count and prevent errors 

logcounts <- log2(countdata + 1)

# getting the log2 counts 

# Check the distribuion of the counts and colour by condition 

conditionCol <- match(sample_information$Condition, c("Null", "Cre")) + 1

# above we have created a colour vector 

png("counts_distribution_boxplot.png")

boxplot(logcounts,
        xlab= "",
        ylab= "Log2(Counts)",
        las = 1,
        col = conditionCol,
        main = "Counts Distribution Boxplot")
abline(h = median(as.matrix(logcounts)), col = "Blue")

dev.off()

# Above, we built a boxplot of the Log2(counts), coloured by condition
# The line is the median of the log2(counts) 
# remember we have to convert the log2(counts) into a matrix, hence the as.matrix
# Line is coloured blue 


#################################################################################################################################

# PRINCIPLE COMPONENT ANALYSIS - PRE DIFFERENTIAL EXPRESSION (SANITY CHECK)

library(ggfortify)

png("PreDifferentialExpression_PCA.png")

rlogcounts <- rlog(countdata)

pcDat <- prcomp(t(rlogcounts))

autoplot(pcDat,
         data = sample_information,
         colour = "Condition",
         shape = FALSE,
         size = 5)

dev.off()

#################################################################################################################################

# CONVERT COUNTS INTO A DESeqDataSet OBJECT

# A DESeqDataSet object is used by DESeq2 to store count data 
# It has a number of slots for storing; counts, sample information, the model design etc for DE

library(DESeq2)

all(sample_information$Sample == colnames(countdata))

design <- as.formula(~ Mouse + Condition) # correcting for batch effect by including in my design - note: always do condition of interest last  

model_matrix <- model.matrix(design, data = sample_information) 

ddsObj <- DESeqDataSetFromMatrix(countData = countdata,
                                 colData = sample_information,
                                 design = design)

ddsObj$Condition <- relevel(ddsObj$Condition, ref = "Null")

# By default, R chooses the Ctrl group by alphabetical order
# Above, we are explicitly telling DESeq that our reference (or Ctrl sample) is Null

ddsObj <- estimateSizeFactors(ddsObj)

sizeFactors <- sizeFactors(ddsObj)

ddsObj_results <- DESeq(ddsObj)

resultsNames(ddsObj_results)                                            

# MDM2 <- results(ddsObj_results, contrast = c("Condition", "Cre", "Null"), alpha = 0.05)

library(apeglm)

MDM2 <- lfcShrink(ddsObj_results, coef = "Condition_Cre_vs_Null", type = "apeglm")

#################################################################################################################################
# PCA post batch correction 

ddsObj$Mouse <- factor(rep(c("1", "2", "3", "4", "5", "6", "7", "8")))

png("PCAplot_pre_batchcorrection.png")

vsd <- varianceStabilizingTransformation(ddsObj) # do a variance stablizing transformation on the model matrix 

plotPCA(vsd, "Mouse")+
  geom_text(aes(label = ddsObj$Sample), vjust = 1.5) +
  ggtitle("MDM2 Null vs Cre")+ 
  coord_fixed(ratio = 2) # this should give you the variance associated with the effect of the batch 

dev.off()

png("PCAplot_post_batchcorrection.png")

assay(vsd) <- limma::removeBatchEffect(assay(vsd), vsd$Mouse) # applying limma's removeBatchEffect 

plotPCA(vsd, intgroup = c("Mouse")) # plotting the residuals 

dev.off()

# Replot the PCAs in ggplot 

# need to find way to replot the PCA before and after batch correction 

#################################################################################################################################

#prep data for barplot input 

# Generate a table from the DESeq2 results 

MDM2_table <- as.data.frame(MDM2) %>%
  rownames_to_column("Gene") %>%
  rename(log2FC = log2FoldChange, FDR = padj)

# save the DESeq2 results table for future reference 

write_tsv(MDM2_table, "/mnt/data/BMOHAMED/TRAX_nov2019/tRAX-master/MDM2/counts_analysis/nocre1_2/collapsed_anticodon/rawcounts_collapsedanticodon_fulland3prime_DESeq2results.txt")

# reformatt the table so it can be used for tRNA barplot generation 

# step 1 ... remove all rows that are non-tRNAs

barplot_input <- slice(MDM2_table, 1:49) # only take the tRNA data 

# step 2 ... convert Gene column into individual AA and anticodon columns etc  

barplot_input <- mutate(barplot_input, Label = Gene) # make a new column called Label 

barplot_input <- separate(barplot_input, Gene, sep = "-", into = c("Isodecoder", "Anticodon")) # seperate the rowname into columns

# step 3 ... remove all rows with NA in the FDR column 

barplot_input <- barplot_input[!is.na(barplot_input$FDR),]

barplot_input <- mutate(barplot_input,
                        wobble = case_when(str_detect(Anticodon, "^A|^T") ~ "AT_wobble",
                                           str_detect(Anticodon, "^G|^C") ~ "GC_wobble"))

barplot_input <- mutate(barplot_input, Label2 = Anticodon)

#################################################################################################################################
# Bar plot of tRNA log fold changes 
#################################################################################################################################

library(ggplot2)
library(grid)
library(gridExtra)

df <- barplot_input

plot_list <- list()

for (iso in c("Ala", "Arg", "Asn", "Asp", "Cys", "Gln", "Glu", "Gly", "His", "Ile", "Leu", "Lys", "Met", "Phe", "Pro", "SeC", "Ser", "Thr", "Trp", "Tyr", "Val", "iMet")) {
  
  df2 <- subset(df, Isodecoder == iso) 
  
  temp_plot <- ggplot(data = df2, aes(x = Label2, y = log2FC, fill = wobble, label = ifelse(FDR < 0.05, "*", "ns")))+
    geom_bar(stat = "identity")+
    geom_text(aes(label = ifelse(FDR < 0.05, "*", paste("p=", round(FDR, 3), sep = "")), 
                  vjust = ifelse(df2$log2FC < 0, 1, -0.2), 
                  fontface = "bold"))+
    scale_y_continuous(expand = expansion(mult = 0.1))+
    scale_fill_manual(values = c("GC_wobble" = "#00BFC4", "AT_wobble" = "#F8766D"))+
    #geom_text(vjust = 0, nudge_y = ifelse(df2$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(df2$FDR < 0.05, 6, 4))+
    xlab("Isoacceptor")+
    ylab(expression(paste(Log[2],"FC",sep="")))+
    #ylim(c(-5, 5))+
    theme_bw()+
    theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
    ggtitle(iso, "MDM2 Null vs Cre in Mouse Liver")
  
    ggsave(filename = paste(iso, "_full3prime_combined_LFCs.png", sep = "_"), bg = "white", width = 15, height = 7, dpi = 700)
  
  plot_list[[iso]] <- temp_plot
  
  print(temp_plot)
}

#################################################################################################################################
# Visulisation of the DESeq2 data 
#################################################################################################################################

# P-VALUES HISTOGRAM SANITY CHECK 

png("pvalues_histogram.png")

hist(MDM2$pvalue)

dev.off()


#################################################################################################################################

# MA-PLOT POST DE

png("MA_plot_NullvsCre.png")

plotMA(MDM2, alpha = 0.05,
       main = "MDM2 Null vs Cre in Mouse Liver")

dev.off()

#################################################################################################################################

# Volvano Plot

library(tidyverse)
library(ggplot2)

df <- as.data.frame(MDM2_table)

df %>%
  mutate(DE = factor(case_when(FDR < 0.05 & log2FC > 0 ~ "Up",
                               FDR < 0.05 & log2FC < 0 ~ "Down",
                               FDR >= 0.05 ~ "NS"), 
                     ordered = T, 
                     levels = c("Up", "Down", "NS"), 
                     labels = c("Up", "Down", "Not Sig"))) %>%
  arrange(FDR) -> new_df

new_df <- na.omit(new_df)

png("NullCre_volcanoplot.png")

ggplot(data = new_df, aes(x = log2FC, y = -log10(FDR), colour = DE))+
  geom_point(size = 1, alpha = 0.5)+
  geom_vline(xintercept = -1.5, linetype = "dashed")+
  geom_vline(xintercept = 1.5, linetype = "dashed")+
  scale_colour_manual(values = c("Up" = "red", "Down" = "blue", "Not Sig" = "grey"))+
  xlim(-8, 8)+
  ylim(0, 310)+
  scale_x_continuous(breaks = seq(-8, 8, by = 2))+
  scale_y_continuous(breaks = seq(0, 300, by = 50))+
  geom_hline(yintercept = 1.30103, linetype = "dashed")+
  ylab("-log10(FDR)")+
  xlab("log2FC")+
  ggtitle("MDM2 Null vs Cre in Mouse Liver")+
  theme_bw()+
  theme(axis.title = element_text(size = 20),
        axis.text = element_text(size = 18),
        legend.title = element_blank(),
        legend.text = element_text(size = 20),
        plot.title = element_text(size = 20, hjust = 0.5),
        panel.grid = element_blank())

dev.off()

#################################################################################################################################

# Interactive StripChart with Glimma

library(Glimma)

group <- sample_information$Condition

de <- as.integer(MDM2$padj <= 0.05 & !is.na(MDM2$padj))

normCounts <- log2(counts(ddsObj))

glXYPlot(
  x = MDM2$log2FoldChange,
  y = -log10(MDM2$pvalue),
  xlab = "log2FC",
  ylab = "-log10(FDR)",
  side.xlab = "Condition",
  side.ylab = "Log(CPM)",
  main = "MDM2 Null vs Cre in Mouse Liver",
  counts = normCounts,
  groups = group,
  sample.cols = conditionCol,
  status = de,
  #anno = shrink_Day4_ConVsCre[, c("GeneID", "Symbol", "Description")],
  folder = "volcano")

#################################################################################################################################

# HEATMAPS

library(ComplexHeatmap)
library(circlize)

# get the top 200 genes
sigGenes <- as.data.frame(MDM2_table) %>% 
  top_n(200, wt=-FDR) %>% 
  pull("Gene")

# filter the data for the top 200 by padj in the LRT test
plotDat <- varianceStabilizingTransformation(ddsObj)[sigGenes,] %>% 
  assay()
z.mat <- t(scale(t(plotDat), center=TRUE, scale=TRUE))

# colour palette
myPalette <- c("red3", "ivory", "blue3")
myRamp = colorRamp2(c(-2, 0, 2), myPalette)

Heatmap(z.mat, name = "z-score",
        col = myRamp,            
        show_row_names = FALSE,
        cluster_columns = FALSE)

# cluster the data and split the tree
hcDat <- hclust(dist(z.mat))
cutGroups <- cutree(hcDat, h=4)

ha1 = HeatmapAnnotation(df = colData(ddsObj)[,c("Condition", "Mouse")])

png("heatmap.png")

Heatmap(z.mat, name = "z-score",
        col = myRamp,
        show_row_name = FALSE,
        cluster_columns = FALSE,
        split=cutGroups,
        rect_gp = gpar(col = "darkgrey", lwd=0.5),
        top_annotation = ha1)

dev.off()

#################################################################################################################################
# Barplot of AT ending tRNAs 
#################################################################################################################################
# Here, I'm filtering all anticodons that begin with either an A or T 
# remember that we are looking at tRNAs and so the anticodon is in reverse complement (i.e. first position)

at_ending <- barplot_input %>%
  filter(str_detect(Anticodon, "^A|^T"))

a_ending <- barplot_input %>%
  filter(str_detect(Anticodon, "^A"))

t_ending <- barplot_input %>%
  filter(str_detect(Anticodon, "^T"))

plot_list <- list()

ggplot(data = at_ending, aes(x = Label2, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(at_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(at_ending$FDR < 0.05, 6, 4))+
  xlab("Isoacceptor")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  #ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("A/T at the anticodon wobble position - MDM2 Null vs Cre in Mouse Liver")

  ggsave(filename = paste("AT_ending_tRNA_LFCs.png"), bg = "white", width = 15, height = 7, dpi = 700)

plot_list[[iso]] <- temp_plot


#################################################################################################################################
# Barplot of GC ending tRNAs 
#################################################################################################################################
# Here, I'm filtering all anticodons that begin with either a G or C 
# remember that we are looking at tRNAs and so the anticodon is in reverse complement (i.e. first position)

gc_ending <- barplot_input %>%
  filter(str_detect(Anticodon, "^G|^C"))

g_ending <- barplot_input %>%
  filter(str_detect(Anticodon, "^G"))

c_ending <- barplot_input %>%
  filter(str_detect(Anticodon, "^C"))

plot_list <- list()

ggplot(data = gc_ending, aes(x = Label2, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(gc_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(gc_ending$FDR < 0.05, 6, 4))+
  xlab("Isoacceptor")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  #ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("G/C at the anticodon wobble position - MDM2 Null vs Cre in Mouse Liver")

  ggsave(filename = paste("GC_ending_tRNA_LFCs.png"), bg = "white", width = 15, height = 7, dpi = 700)

plot_list[[iso]] <- temp_plot


#################################################################################################################################
# plots of the normalized counts for each replicate 
#################################################################################################################################

# normalize the raw counts 

normalized_counts <- seq_data

# divide all the counts by their respective size factors 

normalized_counts <- mapply("/", normalized_counts[intersect(names(normalized_counts), names(sizeFactors))], sizeFactors[intersect(names(normalized_counts), names(sizeFactors))])

# combine the barplot table to get the gene names again 

normalized_counts <- cbind(barplot_input, normalized_counts)

# removing columns that are not needed 

normalized_counts <- select(normalized_counts, -c(3,4,5,6,7,8))

# calculate the mean of the normalized counts and add a new column for each 

normalized_counts <- mutate(normalized_counts, Null_Mean = rowMeans(cbind(normalized_counts$Null1, normalized_counts$Null2, normalized_counts$Null3, normalized_counts$Null4, normalized_counts$Null5)))

normalized_counts <- mutate(normalized_counts, Cre_Mean = rowMeans(cbind(normalized_counts$Cre1, normalized_counts$Cre2, normalized_counts$Cre3)))

# plot scatter graph for each anticodon with the mean as a line going through the scatter 

library(ggplot2)
library(grid)
library(gridExtra)
library(reshape2)
library(tidyr)
library(cowplot)

df3 <- normalized_counts

df3_melt <- tidyr::gather(df3, variable, value, Null1:Cre3)

df3_melt <- mutate(df3_melt, Condition = case_when(variable == "Null1" ~ "Null",
                                                   variable == "Null2" ~ "Null",
                                                   variable == "Null3" ~ "Null",
                                                   variable == "Null4" ~ "Null",
                                                   variable == "Null5" ~ "Null",
                                                   variable == "Cre1" ~ "Cre",
                                                   variable == "Cre2" ~ "Cre",
                                                   variable == "Cre3" ~ "Cre"))


for (iso in c("Ala", "Arg", "Asn", "Asp", "Cys", "Gln", "Glu", "Gly", "His", "Ile", "Leu", "Lys", "Met", "Phe", "Pro", "SeC", "Ser", "Thr", "Trp", "Tyr", "Val", "iMet")) {
  
  df4 <- subset(df3_melt, Isodecoder == iso) 
  
  normalized_counts_graphs <- ggplot(data = df4, aes(x = Label2, y = value, colour = Condition, group = Condition))+
    geom_point(position = position_dodge(width = 0.2))+
    #facet_wrap( ~ Condition)+
    #stat_summary(fun.y = mean, aes(group = 1), geom = "point", colour = "black", shape = 4, size = 5)+
    #ggrepel::geom_text_repel(aes(label = variable), colour = "black", size = 2.5, segment.colour = "grey")+
    xlab("Isoacceptor")+
    ylab("Normalized Counts")+
    theme_bw()+
    theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
    ggtitle(iso, "MDM2 Null vs Cre - Normalized Counts")
  
  df2 <- subset(barplot_input, Isodecoder == iso) 
  
  tRNA_barplot <- ggplot(data = df2, aes(x = Label2, y = log2FC, fill = wobble, label = ifelse(FDR < 0.05, "*", "ns")))+
    geom_bar(stat = "identity")+
    geom_text(aes(label = ifelse(FDR < 0.05, "*", paste("p=", round(FDR, 3), sep = "")), 
                  vjust = ifelse(df2$log2FC < 0, 1, -0.2), 
                  fontface = "bold"))+
    scale_y_continuous(expand = expansion(mult = 0.1))+
    scale_fill_manual(values = c("GC_wobble" = "#00BFC4", "AT_wobble" = "#F8766D"))+
    #geom_text(vjust = 0, nudge_y = ifelse(df2$log2FC < 0, -1, 1), size = ifelse(df2$FDR < 0.05, 6, 4))+
    xlab("Isoacceptor")+
    ylab(expression(paste(Log[2],"FC",sep="")))+
    #ylim(c(-5, 5))+
    theme_bw()+
    theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
    ggtitle(iso, "MDM2 Null vs Cre - tRNA LFCs")
  
  grid_panel <- plot_grid(normalized_counts_graphs, tRNA_barplot, ncol = 1, rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)
  
  save_plot(filename = paste(iso, "panel_tRNAsonly.png", sep = "_"), grid_panel, nrow = 2)    
}

# plots with only the mean normalized counts 

df3_mean <- normalized_counts

df3_melt_mean <- tidyr::gather(df3_mean, variable, value, Null_Mean:Cre_Mean)

df3_melt_mean <- mutate(df3_melt_mean, Condition = case_when(variable == "Null_Mean" ~ "Null - Mean of the normalized counts",
                                                             variable == "Cre_Mean" ~ "Cre - Mean of the normalized counts"))


for (iso in c("Ala", "Arg", "Asn", "Asp", "Cys", "Gln", "Glu", "Gly", "His", "Ile", "Leu", "Lys", "Met", "Phe", "Pro", "SeC", "Ser", "Thr", "Trp", "Tyr", "Val", "iMet")) {
  
  df4_mean <- subset(df3_melt_mean, Isodecoder == iso) 
  
  normalized_counts_graphs_mean <- ggplot(data = df4_mean, aes(x = Label2, y = value, colour = Condition, group = Condition))+
    geom_point(position = position_dodge(width = 0.2))+
    #facet_wrap( ~ Condition)+
    #stat_summary(fun.y = mean, aes(group = 1), geom = "point", colour = "black", shape = 4, size = 5)+
    #ggrepel::geom_text_repel(aes(label = variable), colour = "black", size = 2.5, segment.colour = "grey")+
    xlab("Isoacceptor")+
    ylab("Normalized Counts")+
    theme_bw()+
    theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
    ggtitle(iso, "MDM2 Null vs Cre - Mean Normalized Counts")
  
  df2 <- subset(barplot_input, Isodecoder == iso) 
  
  tRNA_barplot_mean <- ggplot(data = df2, aes(x = Label2, y = log2FC, fill = wobble, label = ifelse(FDR < 0.05, "*", "ns")))+
    geom_bar(stat = "identity")+
    geom_text(aes(label = ifelse(FDR < 0.05, "*", paste("p=", round(FDR, 3), sep = "")), 
                  vjust = ifelse(df2$log2FC < 0, 1, -0.2), 
                  fontface = "bold"))+
    scale_y_continuous(expand = expansion(mult = 0.1))+
    scale_fill_manual(values = c("GC_wobble" = "#00BFC4", "AT_wobble" = "#F8766D"))+
    #geom_text(vjust = 0, nudge_y = ifelse(df2$log2FC < 0, -1, 1), size = ifelse(df2$FDR < 0.05, 6, 4))+
    xlab("Isoacceptor")+
    ylab(expression(paste(Log[2],"FC",sep="")))+
    #ylim(c(-5, 5))+
    theme_bw()+
    theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
    ggtitle(iso, "MDM2 Null vs Cre - tRNA LFCs")
  
  grid_panel <- plot_grid(normalized_counts_graphs_mean, tRNA_barplot_mean, ncol = 1, rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)
  
  save_plot(filename = paste(iso, "panel_tRNAsonly_mean.png", sep = "_"), grid_panel, nrow = 2)    
}



#################################################################################################################################
# plots of the normalized counts for each replicate ... individual plots 
#################################################################################################################################

for (iso in c("Ala", "Arg", "Asn", "Asp", "Cys", "Gln", "Glu", "Gly", "His", "Ile", "Leu", "Lys", "Met", "Phe", "Pro", "SeC", "Ser", "Thr", "Trp", "Tyr", "Val", "iMet")) {
  
  df4 <- subset(df3_melt, Isodecoder == iso) 
  
  normalized_counts_graphs <- ggplot(data = df4, aes(x = Label2, y = value, colour = Condition, group = Condition))+
    geom_point(position = position_dodge(width = 0.2))+
    scale_y_continuous(trans='log2')+
    #facet_wrap( ~ Condition)+
    #stat_summary(fun.y = mean, aes(group = 1), geom = "point", colour = "black", shape = 4, size = 5)+
    #ggrepel::geom_text_repel(aes(label = variable), colour = "black", size = 2.5, segment.colour = "grey")+
    xlab("Isoacceptor")+
    ylab("Normalized Counts")+
    theme_bw()+
    theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
    ggtitle(iso, "MDM2 Null vs Cre - Normalized Counts")
  
    ggsave(filename = paste(iso, "_full3primetRNAsonly_normalizedcounts.png", sep = "_"), bg = "white", width = 15, height = 7, dpi = 700)
}

#################################################################################################################################
# plots of the normalized counts AND LFC depending on wobble position 
#################################################################################################################################

# GC at the wobble position 

gc_ending_figure <- normalized_counts %>%
  filter(str_detect(Anticodon, "^G|^C"))

df_gc_wobble_melt <- tidyr::gather(gc_ending_figure, variable, value, Null1:Cre3) 

df_gc_wobble_melt <- mutate(df_gc_wobble_melt, Condition = case_when(variable == "Null1" ~ "Null",
                                                                     variable == "Null2" ~ "Null",
                                                                     variable == "Null3" ~ "Null",
                                                                     variable == "Null4" ~ "Null",
                                                                     variable == "Null5" ~ "Null",
                                                                     variable == "Cre1" ~ "Cre",
                                                                     variable == "Cre2" ~ "Cre",
                                                                     variable == "Cre3" ~ "Cre"))
#variable == "Fed_Mean" ~ "Fed_Mean",
#variable == "Starved_Mean" ~ "Starved_Mean"))


normalized_counts_wobble_graphs_gc <- ggplot(data = df_gc_wobble_melt, aes(x = Label2, y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  scale_y_continuous(trans='log2')+
  xlab("Isoacceptor")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("G/C at the anticodon wobble position", "MDM2 Null vs Cre - Normalized Counts")

GC_at_wobble <- ggplot(data = gc_ending, aes(x = Label2, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(gc_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(gc_ending$FDR < 0.05, 6, 4))+
  xlab("Isoacceptor")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  #ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("G/C at the anticodon wobble position", "MDM2 Null vs Cre - tRNA LFCs")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_gc, GC_at_wobble, ncol = 1, rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "GC_ending_tRNAs_panel.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)




# AT at the wobble position  

at_ending_figure <- normalized_counts %>%
  filter(str_detect(Anticodon, "^A|^T"))

df_at_wobble_melt <- tidyr::gather(at_ending_figure, variable, value, Null1:Cre3) 

df_at_wobble_melt <- mutate(df_at_wobble_melt, Condition = case_when(variable == "Null1" ~ "Null",
                                                                     variable == "Null2" ~ "Null",
                                                                     variable == "Null3" ~ "Null",
                                                                     variable == "Null4" ~ "Null",
                                                                     variable == "Null5" ~ "Null",
                                                                     variable == "Cre1" ~ "Cre",
                                                                     variable == "Cre2" ~ "Cre",
                                                                     variable == "Cre3" ~ "Cre"))


normalized_counts_wobble_graphs_at <- ggplot(data = df_at_wobble_melt, aes(x = Label2, y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  scale_y_continuous(trans='log2')+
  xlab("Isoacceptor")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("A/T at the anticodon wobble position", "MDM2 Null vs Cre - Normalized Counts")

AT_at_wobble <- ggplot(data = at_ending, aes(x = Label2, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(at_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(at_ending$FDR < 0.05, 6, 4))+
  xlab("Isoacceptor")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  #ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("A/T at the anticodon wobble position", "MDM2 Null vs Cre")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_at, AT_at_wobble, ncol = 1, rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "AT_ending_tRNAs_panel.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)


#################################################################################################################################
# plots of the normalized counts AND LFC depending on wobble position ... only showing the mean normalized count
#################################################################################################################################


# GC at the wobble position 

gc_ending_figure_mean <- normalized_counts %>%
  filter(str_detect(Anticodon, "^G|^C"))

df_gc_wobble_melt_mean <- tidyr::gather(gc_ending_figure_mean, variable, value, Null_Mean:Cre_Mean) 

df_gc_wobble_melt_mean <- mutate(df_gc_wobble_melt_mean, Condition = case_when(variable == "Null_Mean" ~ "Null - Mean of the normalized counts",
                                                                               variable == "Cre_Mean" ~ "Cre - Mean of the normalized counts"))


normalized_counts_wobble_graphs_gc_mean <- ggplot(data = df_gc_wobble_melt_mean, aes(x = Label2, y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  scale_y_continuous(trans='log2')+
  xlab("Isoacceptor")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("G/C at the anticodon wobble position", "MDM2 Null vs Cre - Normalized Counts")

GC_at_wobble_mean <- ggplot(data = gc_ending, aes(x = Label2, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(gc_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(gc_ending$FDR < 0.05, 6, 4))+
  xlab("Isoacceptor")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  #ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("G/C at the anticodon wobble position", "MDM2 Null vs Cre - tRNA LFCs")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_gc_mean, GC_at_wobble_mean, ncol = 1, rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "GC_ending_tRNAs_panel_mean.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)




# AT at the wobble position  

at_ending_figure_mean <- normalized_counts %>%
  filter(str_detect(Anticodon, "^A|^T"))

df_at_wobble_melt_mean <- tidyr::gather(at_ending_figure_mean, variable, value, Null_Mean:Cre_Mean) 

df_at_wobble_melt_mean <- mutate(df_at_wobble_melt_mean, Condition = case_when(variable == "Null_Mean" ~ "Null - Mean of the normalized counts",
                                                                               variable == "Cre_Mean" ~ "Cre - Mean of the normalized counts"))


normalized_counts_wobble_graphs_at_mean <- ggplot(data = df_at_wobble_melt_mean, aes(x = Label2, y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  scale_y_continuous(trans='log2')+
  xlab("Isoacceptor")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("A/T at the anticodon wobble position", "MDM2 Null vs Cre - Normalized Counts")

AT_at_wobble_mean <- ggplot(data = at_ending, aes(x = Label2, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(at_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(at_ending$FDR < 0.05, 6, 4))+
  xlab("Isoacceptor")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  #ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("A/T at the anticodon wobble position", "MDM2 Null vs Cre - tRNA LFCs")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_at_mean, AT_at_wobble_mean, ncol = 1, rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "AT_ending_tRNAs_panel_mean.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)


#################################################################################################################################
# seperating counts into individual Wobble position nucleotides 
#################################################################################################################################

# A at the wobble position  

a_ending_figure <- normalized_counts %>%
  filter(str_detect(Anticodon, "^A"))

df_a_wobble_melt <- tidyr::gather(a_ending_figure, variable, value, Null1:Cre3) 

df_a_wobble_melt <- mutate(df_a_wobble_melt, Condition = case_when(variable == "Null1" ~ "Null",
                                                                   variable == "Null2" ~ "Null",
                                                                   variable == "Null3" ~ "Null",
                                                                   variable == "Null4" ~ "Null",
                                                                   variable == "Null5" ~ "Null",
                                                                   variable == "Cre1" ~ "Cre",
                                                                   variable == "Cre2" ~ "Cre",
                                                                   variable == "Cre3" ~ "Cre"))


normalized_counts_wobble_graphs_a <- ggplot(data = df_a_wobble_melt, aes(x = Label2, y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  scale_y_continuous(trans='log2')+
  xlab("Isoacceptor")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("A at the anticodon wobble position", "MDM2 Null vs Cre - Normalized Counts")

A_at_wobble <- ggplot(data = a_ending, aes(x = Label2, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(a_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(a_ending$FDR < 0.05, 6, 4))+
  xlab("Isoacceptor")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  #ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("A at the anticodon wobble position", "MDM2 Null vs Cre")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_a, A_at_wobble, ncol = 1, rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "A_ending_tRNAs_panel.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)

#################################################################################################################################

# T at the wobble position  

t_ending_figure <- normalized_counts %>%
  filter(str_detect(Anticodon, "^T"))

df_t_wobble_melt <- tidyr::gather(t_ending_figure, variable, value, Null1:Cre3) 

df_t_wobble_melt <- mutate(df_t_wobble_melt, Condition = case_when(variable == "Null1" ~ "Null",
                                                                   variable == "Null2" ~ "Null",
                                                                   variable == "Null3" ~ "Null",
                                                                   variable == "Null4" ~ "Null",
                                                                   variable == "Null5" ~ "Null",
                                                                   variable == "Cre1" ~ "Cre",
                                                                   variable == "Cre2" ~ "Cre",
                                                                   variable == "Cre3" ~ "Cre"))


normalized_counts_wobble_graphs_t <- ggplot(data = df_t_wobble_melt, aes(x = Label2, y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  scale_y_continuous(trans='log2')+
  xlab("Isoacceptor")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("T at the anticodon wobble position", "MDM2 Null vs Cre - Normalized Counts")

T_at_wobble <- ggplot(data = t_ending, aes(x = Label2, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(t_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(t_ending$FDR < 0.05, 6, 4))+
  xlab("Isoacceptor")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  #ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("T at the anticodon wobble position", "MDM2 Null vs Cre")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_t, T_at_wobble, ncol = 1, rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "T_ending_tRNAs_panel.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)

#################################################################################################################################

# G at the wobble position  

g_ending_figure <- normalized_counts %>%
  filter(str_detect(Anticodon, "^G"))

df_g_wobble_melt <- tidyr::gather(g_ending_figure, variable, value, Null1:Cre3) 

df_g_wobble_melt <- mutate(df_g_wobble_melt, Condition = case_when(variable == "Null1" ~ "Null",
                                                                   variable == "Null2" ~ "Null",
                                                                   variable == "Null3" ~ "Null",
                                                                   variable == "Null4" ~ "Null",
                                                                   variable == "Null5" ~ "Null",
                                                                   variable == "Cre1" ~ "Cre",
                                                                   variable == "Cre2" ~ "Cre",
                                                                   variable == "Cre3" ~ "Cre"))


normalized_counts_wobble_graphs_g <- ggplot(data = df_g_wobble_melt, aes(x = Label2, y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  scale_y_continuous(trans='log2')+
  xlab("Isoacceptor")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("G at the anticodon wobble position", "MDM2 Null vs Cre - Normalized Counts")

G_at_wobble <- ggplot(data = g_ending, aes(x = Label2, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(g_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(g_ending$FDR < 0.05, 6, 4))+
  xlab("Isoacceptor")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  #ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("G at the anticodon wobble position", "MDM2 Null vs Cre")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_g, G_at_wobble, ncol = 1, rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "G_ending_tRNAs_panel.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)

#################################################################################################################################

# C at the wobble position  

c_ending_figure <- normalized_counts %>%
  filter(str_detect(Anticodon, "^C"))

df_c_wobble_melt <- tidyr::gather(c_ending_figure, variable, value, Null1:Cre3) 

df_c_wobble_melt <- mutate(df_c_wobble_melt, Condition = case_when(variable == "Null1" ~ "Null",
                                                                   variable == "Null2" ~ "Null",
                                                                   variable == "Null3" ~ "Null",
                                                                   variable == "Null4" ~ "Null",
                                                                   variable == "Null5" ~ "Null",
                                                                   variable == "Cre1" ~ "Cre",
                                                                   variable == "Cre2" ~ "Cre",
                                                                   variable == "Cre3" ~ "Cre"))


normalized_counts_wobble_graphs_c <- ggplot(data = df_c_wobble_melt, aes(x = Label2, y = value, colour = Condition, group = Condition))+
  geom_point(position = position_dodge(width = 0.5))+
  scale_y_continuous(trans='log2')+
  xlab("Isoacceptor")+
  ylab("Normalized Counts")+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(size = 12))+
  ggtitle("C at the anticodon wobble position", "MDM2 Null vs Cre - Normalized Counts")

C_at_wobble <- ggplot(data = c_ending, aes(x = Label2, y = log2FC, fill = Isodecoder, label = ifelse(FDR < 0.05, "*", "ns")))+
  geom_bar(stat = "identity")+
  geom_text(vjust = 0, nudge_y = ifelse(c_ending$log2FC < 0, -0.5, 0.5), fontface = "bold", size = ifelse(c_ending$FDR < 0.05, 6, 4))+
  xlab("Isoacceptor")+
  ylab(expression(paste(Log[2],"FC",sep="")))+
  #ylim(c(-4, 4))+
  theme_bw()+
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1), plot.margin = margin(0.5,0.5,0.5,2, "cm"), axis.title = element_text(face = "bold", size = 12))+
  ggtitle("C at the anticodon wobble position", "MDM2 Null vs Cre")

grid_panel <- plot_grid(normalized_counts_wobble_graphs_c, C_at_wobble, ncol = 1, rel_widths = c(1,1), align = "vh", vjust = 1, scale = 1)

save_plot(filename = "C_ending_tRNAs_panel.png", grid_panel, nrow = 2, base_width = 15, base_height = 7)

#################################################################################################################################
# Place all A/T/G/C ending wobble positions into one panel
#################################################################################################################################

grid_panel_ATCG_wobble <- plot_grid(A_at_wobble, T_at_wobble, G_at_wobble, C_at_wobble, ncol = 2)

save_plot(filename = "ATGC_ending_tRNAs_panel.png", grid_panel_ATCG_wobble, nrow = 2, base_width = 15, base_height = 7)
































